src/decomp_2d_nccl.f90

```diff
@@ -0,0 +1,309 @@
+             
+   public :: decomp_2d_nccl_init, &
+             decomp_2d_nccl_fin, &
+             decomp_2d_nccl_send_recv_col, &
+             decomp_2d_nccl_send_recv_row
+
+   interface decomp_2d_nccl_send_recv_col
+      module procedure decomp_2d_nccl_send_recv_real_col 
+      module procedure decomp_2d_nccl_send_recv_cmplx_col 
+   end interface decomp_2d_nccl_send_recv_col
+
+   interface decomp_2d_nccl_send_recv_row
+      module procedure decomp_2d_nccl_send_recv_real_row
+      module procedure decomp_2d_nccl_send_recv_cmplx_row 
+   end interface decomp_2d_nccl_send_recv_row
+
+contains
+   !
+   ! init of the arrays
+   ! 
+   subroutine decomp_2d_nccl_init(COMM_COL,COMM_ROW)
+      
+      implicit none
+
+      integer, intent(in) :: COMM_COL, COMM_ROW
+      integer             :: ierror
+      integer :: cuda_stat
+      type(ncclResult) :: nccl_stat
+
+      call MPI_COMM_RANK(COMM_COL, col_rank, ierror)
+      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_COMM_RANK")
+      call MPI_COMM_RANK(COMM_ROW, row_rank, ierror)
+      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_COMM_RANK")
+      call MPI_COMM_SIZE(COMM_COL, col_comm_size, ierror)
+      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_COMM_SIZE")
+      call MPI_COMM_SIZE(COMM_ROW, row_comm_size, ierror)
+      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_COMM_SIZE")
+
+      allocate (local_to_global_col(col_comm_size), local_to_global_row(row_comm_size))
+
+
+      local_to_global_col(:) = 0
+      local_to_global_row(:) = 0
+      local_to_global_col(col_rank + 1) = nrank
+      local_to_global_row(row_rank + 1) = nrank
+
+      call mpi_allreduce(MPI_IN_PLACE, local_to_global_col, col_comm_size, MPI_INTEGER, MPI_SUM, COMM_COL, ierror)
+      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_ALLREDUCE")
+      call mpi_allreduce(MPI_IN_PLACE, local_to_global_row, row_comm_size, MPI_INTEGER, MPI_SUM, COMM_ROW, ierror)
+      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_ALLREDUCE")
+
+      if (nrank == 0) then
+         nccl_stat = ncclGetUniqueId(nccl_uid_2decomp)
+         if (nccl_stat /= ncclSuccess) call decomp_2d_abort(__FILE__, __LINE__, nccl_stat, "ncclGetUniqueId")
+      end if
+      call MPI_Bcast(nccl_uid_2decomp, int(sizeof(ncclUniqueId)), MPI_BYTE, 0, decomp_2d_comm, ierror)
+      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_BCAST")
+
+      nccl_stat = ncclCommInitRank(nccl_comm_2decomp, nproc, nccl_uid_2decomp, nrank)
+      if (nccl_stat /= ncclSuccess) call decomp_2d_abort(__FILE__, __LINE__, nccl_stat, "ncclCommInitRank")
+      cuda_stat = cudaStreamCreate(cuda_stream_2decomp)
+      if (cuda_stat /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "cudaStreamCreate")
+   
+   end subroutine decomp_2d_nccl_init
+   !
+   ! init of the arrays
```

Comment:
?

Reply:
In this a question on the comment?

Reply:
The comment for decomp_2d_ccl_fin should not be "init of the arrays"